util.go

```diff
@@ -126,3 +129,58 @@ func NetInterval(cidr string) (net.IP, net.IP, error) {
 
 	return first, nextIP(last), nil
 }
+
+// cidrString returns an IPnet given a network address and CIDR mask
+func cidrString(address net.IP, cidr int) (*net.IPNet, error) {
+	_, out, err := net.ParseCIDR(fmt.Sprintf("%s/%d", address, cidr))
+	if err != nil {
+		return nil, err
+	}
+
+	return out, err
+}
+
+// NetFromRange returns a CIDR IP network given a start and end address
+func NetFromRange(first []byte, last []byte) (*net.IPNet, error) {
+	ip1 := net.IP(first)
+	ip2 := net.IP(last)
+
+	maxLen := 32
+	isIpv6 := ip1.To4() == nil
+
+	if isIpv6 && ip2.To4() != nil {
+		return nil, errors.New("Cannot mix IPv4 and IPv6.")
+	}
+
+	if isIpv6 {
+		maxLen = 128
+	}
+
+	for l := maxLen; l >= 0; l-- {
+		cidrmask := net.CIDRMask(l, maxLen)
+		ipmask := ip2.Mask(cidrmask)
+		ipnet := net.IPNet{
+			IP:   ipmask,
+			Mask: cidrmask,
+		}
+
+		if ipnet.Contains(ip1) {
+			return cidrString(ipmask, l)
+		}
+	}
+
+	return nil, errors.New("Failed to construct network from range.")
+}
+
+// NetFromNetInterval returns a CIDR IP network given a start and end address as found in intervals.
+// This is similar to NetFromRange, but subtracts one address from the end of the range.
+func NetFromIntervalRange(first []byte, last []byte) (out *net.IPNet, err error) {
+	ip2, ok := netip.AddrFromSlice(last)
+	if !ok {
+		return nil, errors.New("Failed to construct slice from network.")
+	}
+
+	previous := ip2.Prev()
```

Comment:
What happens if there is no previous IP ?

Reply:
From my understanding it will just return the address itself, for example calling `Prev()` on `0.0.0.0`, will yield `0.0.0.0`.

Reply:
This is the `IntervalOpen` scenario. Consider a CIDR like `128.0.0.0/1`. In that instance, the user would have to call `NetIntervalRange` by setting `last` into an empty IP or nil. 

I'd suggest an empty IP instead of a nil value for consistency with `NetInterval` but I have no strong opinions about this.

Reply:
Thanks for the explanation!